docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,263 @@
+
+## Goals
+
+Safety is a top priority for the Swift programming language. C++ often puts safety in the hands of the programmer. Nullability, memory safety, lifetimes, and mutability are all the programmers responsibility in C++. Swift has the opposite default behavior in many of these cases. Simply using C++ APIs should not completely destroy the memory safety, nullability check, etc. that Swift tries to offer. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the nessisary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer, and even make those APIs unavailable if there's no reasonable path to a sufficently safe interface.
+
+C++ interoperability should strive to have good diagnostics. When an API cannot be imported, this needs to be communicated to the programmer. The compiler should produce a clear error any time an API that could not be imported is used, and should strive to guide the programmer towards an API interface that can be imported (for example, by suggesting annotations). The compiler should warn users about incorrect API usage, and potentially slow or unsafe patterns. C++ often has complex type names, especially for template specializations, this is an area where the Swift compiler can do better than C++ by showing shorter and more helpful diagnostics, right at the call-site. When the Swift compiler can present better diagnostics than Clang, it absolutely should. 
+
+C++ provides tools to create high-performance APIs. C++ interoperability should embrace this. Interop should not be a significant source of overhead and performance concerns should not be a reason to continue using C++ to call C++ APIs rather than Swift.
```

Comment:
```suggestion
C++ provides tools to create high-performance APIs. C++ interoperability in Swift should embrace this. Interop should not be a significant source of overhead, and performance concerns should not be a reason to continue using C++ to call C++ APIs rather than Swift.
```